articles/151_node_to_participant_mapping.md

```diff
@@ -0,0 +1,241 @@
+---
+layout: default
+title: Node to Participant mapping
+permalink: articles/node_to_participant_mapping.html
+abstract: This article analyzes the performance implications of enforcings a one-to-one mapping between ROS nodes and DDS participants, and propose alternative approaches.
+author: '[Ivan Paunovic](https://github.com/ivanpauno)'
+published: true
+categories: Middleware
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Introduction
+
+### What is a `Node`?
+
+In ROS, a `Node` is an entity used to group other entities.
+For example: `Publishers`, `Subscriptions`, `Services`, `Clients`.
+`Nodes` ease organization and code reuse, as they can be composed and launched in different ways.
+
+### What is a `Domain Participant`?
+
+A `Domain Participant` is a type of DDS entity.
+`Participants` also group other entities, like `Publishers`, `Subscribers`, `Data Writters`, `Data Readers`, etc.
+But participants do more than that:
+
+- Each `Participant` does discovery by its own.
+  Creating more than one `Participant` increases cpu usage and network IO load.
+- Each `Participant` keeps track of other `Domain Participants` and DDS entities.
+  Using more than one will duplicate that data.
+- Each `Participant` may create multiple threads for event handling, discovery, etc.
+  The number of threads created per participant depend on the DDS vendor (e.g.: [connext](https://community.rti.com/best-practices/create-few-domainparticipants-possible)).
+
+For those reasons, `Participants` are heavyweight.
+
+### Current status
+
+There is a one-to-one mapping between `Nodes` and `DDS Participants`.
+This simplified the design, as `DDS Participants` provide the same organization that a `Node` needs.
+The drawback of this approach, is that with an increasing number of nodes the overhead also increases.
+Furthermore, the maximum number of `Domain participants` is rather small.
+For example, [RTI connext](https://community.rti.com/kb/what-maximum-number-participants-domain) is limited to 120 participants per domain.
+
+## Proposal
+
+The goal of this proposal is to improve overall performance by avoiding the creation of one `Domain Participant` per `Node`.
+API changes will be avoided, if possible.
+
+### What is a participant mapped to?
+
+There are two main alternatives:
+- One participant per process.
+- One participant per context.
+
+The second approach allows more flexibility.
+Considering that by default there's only one context per process, it wouldn't affect the case where each node runs in its own process.
+And even in the case that multiple nodes are running in a single process it allows to group them into different context - ranging from a separate context for each node, over grouping a few nodes in the same context, to using a single context for all nodes.
+
+### What is a Node now?
```

Comment:
With the exception of the last sentence, I feel like this section is just a more verbose version of the introduction section about nodes. I think it may make sense to merge this with the introduction. The last sentence can be appended to the previous section as part of the consequences for choosing one participant per context, e.g.

> A consequence of removing the 1-to-1 mapping between nodes and participants is each `Participant` should store all the information needed about its nodes, and communicate it other `Participants`.